wpilibc/src/main/native/include/frc/event/BooleanEvent.h

```diff
@@ -131,5 +131,6 @@ class BooleanEvent {
  private:
   EventLoop* m_loop;
   std::function<bool()> m_condition;
+  bool m_state = false;  // A programmer's worst nightmare.
```

Comment:
References to this will become invalid (dangling) once the BooleanEvent is destroyed. Perhaps `shared_ptr<bool>`?
Also, the composition decorators will need to be adapted for this change.

Reply:
Should all of the other uses of `m_condition()/m_condition.getAsBoolean()` (in `getAsBoolean()`, `rising()`, `falling()`, `debounce()`, `negate()`, `and()`, and `or()`) also be replaced with `m_state` or should only `and()` and `or()` be changed?

Reply:
Probably should replace all uses. The idea is that if the signal is true at loop poll, then it should stay true for the rest of the poll and not switch.

Reply:
Why can't we access `m_state` from the other `BooleanEvent`? Access control is done by class, not by instance.

We should define what the desired behavior is if `condition()` returns different values within a single poll of an event loop (which would also apply to `Trigger`)- At present, there's a mix, since `ifHigh()` currently reuses the same value across a poll, but each composition will use a different value.

Reply:
Huh. I learned something new today. In fairness, Java used a `BooleanSupplier` and C++ used a `std::function<bool()>`. I switched them to `BooleanEvent` and I can in fact access m_state.

The problem really does come down to handling `condition()` returning a different value at different times. Defining the behavior of that would be helpful. IMHO, I think that if a condition returns different values within a single poll, the first value it gets should apply to the rest of the poll. Importantly, either all or none of the actions should be executed, having only some of them execute (current behavior) is confusing.

Reply:
We should probably avoid capturing the BooleanEvent itself, so it can be destroyed / kept temporarily -- access the `AtomicBoolean`/`shared_ptr<bool>` directly (and I hope compilers will be smart).

Reply:
> In fairness, Java used a BooleanSupplier and C++ used a std::function<bool()>. I switched them to BooleanEvent and I can in fact access m_state.

I guess this is what I get for rushing a comment without double-checking the code- I didn't realize the parameters weren't `BooleanEvent`. Also don't miss Starlight's comment- The callback's already there (though it should be fixed), just make sure to only capture the shared pointer and not the whole object.